Let user daemon run CLI commands using a PTY device

thallgren opened the pull request:
## Description

This commit adds a pseudo TTY (PTY) when running CLI commands from the
user daemon, so that the commands have the same behaviour as they do
when running in a terminal.

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
Let user daemon run CLI commands using a PTY device

This commit adds a pseudo TTY (PTY) when running CLI commands from the
user daemon, so that the commands have the same behaviour as they do
when running in a terminal.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

raphaelreyna commented on pkg/client/userd/grpc.go:
```diff
@@ -498,22 +501,56 @@ func (s *Service) ListCommands(ctx context.Context, _ *empty.Empty) (groups *rpc
+	} else {
+		// Use a PTY on unix systems.
+		var err error
+		if wr, rd, err = pty.Open(); err != nil {
```
should we log this error?
another option could be to open the pty and doing an error check before launching this func as goroutine, this way we also avoid the external process exiting because this silently failed.

thallgren replied:
Good catch! I meant for this error to be returned and propagated to the user.

raphaelreyna reviewed (changes requested):
lgtm just a minor change request regarding an error

Commit:
Don't use a PTY unless it's needed

A PTY is only needed when the command is an `telepresence intercept`
with extra arguments after `--`. Also, for `--docker-run`, the PTY is
only needed if `-it` is used.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

raphaelreyna commented on pkg/client/userd/grpc.go:
```diff
@@ -498,22 +502,91 @@ func (s *Service) ListCommands(ctx context.Context, _ *empty.Empty) (groups *rpc
+			if !hasT {
+				hasT = strings.ContainsRune(arg, 't')
+			}
+			if hasI && hasT {
```
So I think we only need `if hasT {}` here since `-t` / `--tty` alone is used to request a pty for the container, `-i`/`--interactive` is specific to stdin but we should still support streaming commands.

thallgren replied:
We actually don't need a PTY unless `-i` is used. Using only `-t` works fine without a PTY. The long forms `--tty` and `--interactive` must be supported though. I'll fix that.

thallgren replied:
There's a good text regarding `-i` and `-t` here: https://stackoverflow.com/a/54254380/328043

raphaelreyna commented on pkg/client/userd/grpc.go:
```diff
@@ -617,6 +685,13 @@ func (s *Service) RunCommand(cmdStream rpc.Connector_RunCommandServer) (err erro
 		}
 
+		var rd io.Reader
+		if ctx, rd, cmdErr = stdinPump(ctx, cmdStream, needsPTY(cmd)); cmdErr != nil {
```
Do we want to log this error or make it known to the user somehow?

thallgren replied:
It gets known to the user. It's assigned to `cmdErr` which is handled in the defer function that sends the final error.

Commit:
Tidy up go.mod a bit

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Clarify when PTY is needed and make long options for -i and -t work

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
